File: packages/client-sdk/src/services/admin/bulk.ts
/**
 * Admin Bulk Operations service
 */

import type { HttpClient } from '../../core/http';
import type {
  BulkCreateUsersRequest,
  BulkUpdateUsersRequest,
  BulkDeleteUsersRequest,
  BulkAssignRolesRequest,
  BulkOperationResult,
} from '../../types/admin';
import { BaseService } from '../base';

/** Admin Bulk Operations service for bulk user management */
export class AdminBulkService extends BaseService {
  constructor(http: HttpClient) {
    super(http);
  }

  /**
   * Bulk create users
   * @param data Bulk create request
   * @returns Operation result
   */
  async bulkCreateUsers(data: BulkCreateUsersRequest): Promise<BulkOperationResult> {
    const response = await this.http.post<BulkOperationResult>('/api/admin/users/bulk-create', data);
    return response.data;
  }

  /**
   * Bulk update users
   * @param data Bulk update request
   * @returns Operation result
   */
  async bulkUpdateUsers(data: BulkUpdateUsersRequest): Promise<BulkOperationResult> {
    const response = await this.http.put<BulkOperationResult>('/api/admin/users/bulk-update', data);
    return response.data;
  }

  /**
   * Bulk delete users
   * @param data Bulk delete request
   * @returns Operation result
   */
  async bulkDeleteUsers(data: BulkDeleteUsersRequest): Promise<BulkOperationResult> {
    const response = await this.http.post<BulkOperationResult>('/api/admin/users/bulk-delete', data);
    return response.data;
  }

  /**
   * Bulk assign roles to users
   * @param data Bulk assign roles request
   * @returns Operation result
   */
  async bulkAssignRoles(data: BulkAssignRolesRequest): Promise<BulkOperationResult> {
    const response = await this.http.post<BulkOperationResult>('/api/admin/users/bulk-assign-roles', data);
    return response.data;
  }
}

File: backend/internal/models/bulk.go
package models

import (
	"github.com/google/uuid"
)

// BulkCreateUsersRequest represents a request to create multiple users
type BulkCreateUsersRequest struct {
	Users []BulkUserCreate `json:"users" binding:"required"`
}

// BulkUserCreate represents a single user to create in bulk operation
type BulkUserCreate struct {
	Email         string `json:"email" binding:"required"`
	Username      string `json:"username" binding:"required"`
	FullName      string `json:"full_name,omitempty"`
	Password      string `json:"password" binding:"required"`
	IsActive      bool   `json:"is_active,omitempty"`
	EmailVerified bool   `json:"email_verified,omitempty"`
}

// BulkUpdateUsersRequest represents a request to update multiple users
type BulkUpdateUsersRequest struct {
	Users []BulkUserUpdate `json:"users" binding:"required"`
}

// BulkUserUpdate represents a single user to update in bulk operation
type BulkUserUpdate struct {
	ID       uuid.UUID `json:"id" binding:"required"`
	Email    *string   `json:"email,omitempty"`
	Username *string   `json:"username,omitempty"`
	FullName *string   `json:"full_name,omitempty"`
	IsActive *bool     `json:"is_active,omitempty"`
}

// BulkDeleteUsersRequest represents a request to delete multiple users
type BulkDeleteUsersRequest struct {
	UserIDs []uuid.UUID `json:"user_ids" binding:"required"`
}

// BulkAssignRolesRequest represents a request to assign roles to multiple users
type BulkAssignRolesRequest struct {
	UserIDs []uuid.UUID `json:"user_ids" binding:"required"`
	RoleIDs []uuid.UUID `json:"role_ids" binding:"required"`
}

// BulkOperationResult represents the result of a bulk operation
type BulkOperationResult struct {
	Total   int                       `json:"total" example:"100"`
	Success int                       `json:"success" example:"95"`
	Failed  int                       `json:"failed" example:"5"`
	Errors  []BulkOperationError      `json:"errors,omitempty"`
	Results []BulkOperationItemResult `json:"results,omitempty"`
}

// BulkOperationError represents an error for a specific item in bulk operation
type BulkOperationError struct {
	Index   int    `json:"index" example:"5"`
	ID      string `json:"id,omitempty" example:"123e4567-e89b-12d3-a456-426614174000"`
	Email   string `json:"email,omitempty" example:"user@example.com"`
	Message string `json:"message" example:"User already exists"`
}

// BulkOperationItemResult represents the result for a single item
type BulkOperationItemResult struct {
	Index   int       `json:"index" example:"0"`
	ID      uuid.UUID `json:"id" example:"123e4567-e89b-12d3-a456-426614174000"`
	Email   string    `json:"email" example:"user@example.com"`
	Success bool      `json:"success" example:"true"`
	Message string    `json:"message,omitempty" example:"User created successfully"`
}
